core/test/network-records-to-devtools-log.js

```diff
@@ -40,20 +46,181 @@ function headersArrayToHeadersDict(headersArray = []) {
+  } = networkRecord;
+
+  const requestTime = timeDefined(requestTimeS) ? requestTimeS * 1000 : undefined;
+  const absoluteTimes = {rendererStartTime, startTime, requestTime, responseReceivedTime, endTime};
+  assertTimingIncreases(absoluteTimes);
+
+  // `requestTime` and `startTime` must be equal if both are defined.
+  if (timeDefined(startTime) && timeDefined(requestTime)) {
+    assert.equal(startTime, requestTime, `'startTime' (${startTime}) is not equal to 'timing.requestTime' (${requestTimeS} seconds) in test network record`); // eslint-disable-line max-len
+  }
+
+  // Start of request + relative timing must be <= responseReceivedTime and endTime.
+  const startTimes = [rendererStartTime, startTime, requestTime];
+  let maxStart = Math.max(...startTimes.filter(timeDefined));
+  maxStart = Number.isFinite(maxStart) ? maxStart : redirectResponseTimestamp; // Use redirectResponseTimestamp only as fallback.
+  if (timeDefined(maxStart)) {
+    if (timeDefined(responseReceivedTime)) {
+      assert(maxStart + relativeTimingMax <= responseReceivedTime, `request start (${maxStart}) plus relative 'timing' value (${relativeTimingMax}) exceeds 'responseReceivedTime' (${responseReceivedTime}) in test network record`); // eslint-disable-line max-len
+    }
+    if (timeDefined(endTime)) {
+      assert(maxStart + relativeTimingMax <= endTime, `request start (${maxStart}) plus relative 'timing' value (${relativeTimingMax}) exceeds 'endTime' (${endTime}) in test network record`); // eslint-disable-line max-len
+    }
+  }
+
+  // If all are defined, requestTime + receiveHeadersEnd === responseReceivedTime.
+  const {receiveHeadersEnd} = networkRecord.timing ?? {};
+  const start = timeDefined(startTime) ? startTime : requestTime;
+  if (timeDefined(start) && timeDefined(receiveHeadersEnd) && timeDefined(responseReceivedTime)) {
+    assert.equal(start + receiveHeadersEnd, responseReceivedTime, `request start (${start}) plus 'receiveHeadersEnd' (${receiveHeadersEnd}) does not equal 'responseReceivedTime' (${responseReceivedTime}) in test network record`); // eslint-disable-line max-len
+  }
+
+  return {
+    redirectResponseTimestamp,
+    rendererStartTime,
+    startTime,
+    requestTime,
+    receiveHeadersEnd,
+    responseReceivedTime,
+    endTime,
+    relativeTimingMax,
+  };
+}
+
+/**
+ * Takes the partial network timing and fills in the missing values so at least
+ * time is linear (if not realistic). The main timing properties on
+ * `NetworkRequest` and `requestTime` and `receiveHeadersEnd` on
+ * `NetworkRequest['timing']` will be computed. The other `timing` properties
+ * will not be automatically provided, but will be copied if provided, so they
+ * are up to the caller to order correctly. If no absolute times are provided,
+ * starts at `defaultStart`, finishes receiving headers `defaultTimingOffset` ms
+ * later, then endTime is `defaultTimingOffset` ms after that.
```

Comment:
```suggestion
 * are up to the caller to order correctly.
 * If no absolute times are provided: starts at `defaultStart`, finishes receiving
 * headers `defaultTimingOffset` ms later, then endTime is `defaultTimingOffset` ms after that.
```

I'm having some trouble parsing this sentence. Trivial edit above, but hopefully you can improve further.